### ISSUE TITLE
Dynamically changing bucket configuration

### ISSUE THREAD

**Issue by Nevaan** [vladimir-bukhtoyarov/bucket4j#133]
Hi,

I have use case like this: I need to rate limit calls to couple AP The calls are performed in cron-scheduled jobs couple of times a day. I am using Hazelcast as token datastore to keep it synchronized between app nodes, and initially bucket is configured with limits:
- 10 req/s
- 500 req/min
- 10000 req/H

I would like to edit these limits in the runtime, so I keep the configs in another config management microservice. I pull the configs at the start of my job and update the bucket. Everything seems ok as long as I tinkered with the same amount of limits (like I change limit from 500 to 300 req/min), but changing the list of 3 elements to 1 another (for example 300/day) ends up with IncompatibleConfigurationException. In the source code I can see the validation depends on the count of limits that are already set. 

My question is, is there a way to achieve what I want? I couldn't find a Bucket4j api to remove bucket and replace it with new, with same key and different configuration. What I figured is that maybe I can remove it from Hazelcast by myself, but I am not sure if its safe (I have concerns, because my requirement for the feature are to keep application wait for token retrieval without timeout, so its blocking on BlockingBucket consume method)

**Comment by Nevaan** [comment]
OK, it seems that removing it manually is not the best option as it obviously resets already used tokens.

So, is there a reason why I have to replace configuration with exactly same amount of limits? I thought its a matter of refill period, but nothing prevents me from changing single limit of 1 req/hour to 1req/s. Problem occurs for example if I change single limit to multiple on existing bucket

**Comment by vladimir-bukhtoyarov** [comment]
Hello @Nevaan 

Following facts worth to mention about the current implementation of configuration replacing:
* Bucket counts consumed tokens independently per each limit.
* During the configuration change bucket does copying of consumed tokens per each limit (limit[0]->limit[0], limit[1]->limit[1], limit[N]->limit[N]). Where consumed tokens are the difference between capacity and current amount.
* Bucket restricts the changing of limits count because in this case it is not known how to copy current tokens from previous limits. Undefined behavior is the thing that I always try to avoid.

So, you are right. Removing and recreating bucket it is only way to go. I agree that it is an awkward way, and requires two network requests. 

If you really want to totally forget about previously consumed tokens, then it is not hard for me to extend the API by something that requires only one network hop:
``` java
/**
 * Forgot all previous state and reinitializes the bucket with the new configuration. All information about previously 
 * consumed tokens will be lost.
 */
void reset(BucketConfiguration newConfiguration);
```

But... if you want to avoid missing the information about currently available tokens per each limit then it needs to think about how does API should look.

**Comment by Nevaan** [comment]
> During the configuration change bucket does copying of consumed tokens per each limit (limit[0]->limit[0], limit[1]-limit[1]... limit[N]->limit[N]). Where consumed tokens are the difference between capacity and current amount

Oh, does it mean that if initially I have config in given order ([ 10R/second, 500R/minute, 10000/hour ] ) and lets say I make the only change of reordering ([ 10000R/hour, 500R/minute, 10R/second ] ) it will cause bugs?I mean, tokens consumed initially by hour limit will now be assigned to the seconds limit? In the other words, it is only safe to pass limit configuration in the same order and in the same amount of limits?

> Bucket restricts the changing of limits count because in this case it is not known how to copy current tokens from previous limits. Undefined behavior is the thing that I always try to avoid.

This is a behaviour caused by the fact, information about consumed tokens is an array? I think I saw in code that information about refilling is inside of List, while state of capacity & currently available tokens is array ( I guess it was represented by triples, each limit is 3 elements of array).

**Comment by vladimir-bukhtoyarov** [comment]
> Oh, does it mean that if initially I have config in given order ([ 10R/second, 500R/minute, 10000/hour ] ) and lets say I make the only change of reordering ([ 10000R/hour, 500R/minute, 10R/second ] ) it will cause bugs?I

Yes. It will cause an undesired situation. Looks I needed to add more documentation.

> I mean, tokens consumed initially by the hour limit will now be assigned to the seconds limit?

I was a little bit inaccurate in the previous comment. The state of the bucket is described by this field [BucketState#sateData](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/BucketState.java#L36). Logically, it is to a two-dimensional array, which has as many rowa as many bandwidths were configured for the bucket. Where each row stores:
* Last refill date for this limit. Because different limits may have different refill types (see for example greedy and internally refills) each limit needs to remember its own refill timestamp independently from other limits.
* Current amount of available tokens for this limit.
* Computational correction. It is just a boring field that was involved because the formula  ```(t' - t)  * (capacity / period)``` produces the rest of the division when we calculate it in integer numbers.

But physically it was implemented as one-dimension array because of performance considerations. Bucket with new configuration inherits the previous state, with one exclusion: available tokens cannot be greater than new capacity, see this test https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/test/java/io/github/bucket4j/ConfigurationReplacementSpecification.groovy#L97

**Comment by Nevaan** [comment]
Thank you for clarification. For now I think I will just give up with changing the configs dynamically. 

I think it would be worth to consider changing `IncompatibleConfigurationException` to be checked exception - to be honest I discovered this problem accidentally. It would be great hint if I saw it immediately. 

Besides adding or removing bandwidths, current API is quite misleading and IMO now user can cause impredictible behavior. 

Despite all my complaining, I just wanted to say I really think that this is awesome library - very good job 👌 Your responsiveness and support is crazy 😁 .I did not faced any other issues while implementing my feature with Bucket4j.

**Comment by vladimir-bukhtoyarov** [comment]
> For now I think I will just give up with changing the configs dynamically.

Btw, I have proposed an additional method ```reset``` a few comments above. What do you think about this proposal? New method at least replaces two network requests by one.

> I think it would be worth to consider changing IncompatibleConfigurationException to be checked exception - to be honest I discovered this problem accidentally. It would be great hint if I saw it immediately.

Currently, it is not possible to mark IncompatibleConfigurationException as checked, because it is nob-backward compatible change. But it is possible for the next major release ```5.0```

> Besides adding or removing bandwidths, current API is quite misleading and IMO now user can cause impredictible behavior.

I agree, but dynamic configuration replacing is harder than it looks, because of the previous state:
* There are no problems if we want to forget about previously consumed tokens. If you agree that changing of configuration for an empty bucket will restore all tokens then you are lucky.
* But it is a tricky problem when we expect that previous consumption should take effect to the bucket with new configuration.

**Comment by Nevaan** [comment]
> Btw, I have proposed an additional method reset a few comments above. What do you think about this proposal? New method at least replaces two network requests by one.

The point is I don't want to lose tokens in as many scenarios as possible. I do not know how much this is possible, besides this is probably quite task-specific requirement.

> Currently, it is not possible to mark IncompatibleConfigurationException as checked, because it is nob-backward compatible change. But it is possible for the next major release 5.0

👍. Do you agree with my opinion?

> I agree, but dynamic configuration replacing is harder than it looks
I thought so - the most important for me was ensuring what I can expect by library. I understand this is not trivial to reset configuration of specific bandwidth. But, how about extending this API in this way:
- add additional entry to stateData array - some kind of UUID
- add method to retrieve information about bandwidths - with the uuid from above
- add method to change single bandwidth (called with uuid)

I'm not sure if it would help me in my case, bu would let user manage configuration in more flexible manner. Maybe in the future it will be possible to add/remove existing limits :)

**Comment by vladimir-bukhtoyarov** [comment]
> Do you agree with my opinion?
I thought in this way a day ago, but now I prefer to do the refactoring:
* Associate a string identifier with each bandwidth. It will be null by default. Semantically it will be everything what user want, it may be UUID or something else which can be used to distinguish the limits.
* This identifier will be stored outside stateArray. [Bandwidth class](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/Bandwidth.java#L69)  by itself will store it, together with other parameters which are already stored there. It will be possible to configure ID via ```withId``` method.
* Change the semantic of replacing configuration method in the following way:
  * Remove IncompatibleConfigurationException
  * Per-limit state will be inherited by limit id. If some limit in the new configuration does not have a paired limit in the previous configuration, the new limit will be just initialized according to its parameters. That is why we can eliminate IncompatibleConfigurationException
  * ```null``` identifier will not participate in the comparison, in other words ```null`` will not be equivalent to ```null```, with one exception: if both new and previous configuration has only one limit with id ```null``` then these two limits will be treated as equal.
* If you really want to have a method for replacing the configuration of a single limit by its id, then I can add it, it will not be hard. I suppose that this method must throw ```LimitNotFoundException``` if there is no limit with this id.
* As planed, I am going to add ```reset``` method which was described a few messages above in order to cover the case when user wants to forget about previous state after configuration change. 

@Nevaan what do you think about it?

**Comment by vladimir-bukhtoyarov** [comment]
Has been released under the version ```6.0.0```.  Feel free to as any questions according to the new API.
